source/common/router/scoped_rds.cc

```diff
@@ -173,20 +175,23 @@ bool ScopedRdsConfigSubscription::addOrUpdateScopes(
       route_provider_by_scope_.insert({scope_name, std::move(rds_config_provider_helper)});
       scope_name_by_hash_[scoped_route_info->scopeKey().hash()] = scoped_route_info->scopeName();
       scoped_route_map_[scoped_route_info->scopeName()] = scoped_route_info;
-      applyConfigUpdate([scoped_route_info](ConfigProvider::ConfigConstSharedPtr config)
-                            -> ConfigProvider::ConfigConstSharedPtr {
-        auto* thread_local_scoped_config =
-            const_cast<ScopedConfigImpl*>(static_cast<const ScopedConfigImpl*>(config.get()));
-        thread_local_scoped_config->addOrUpdateRoutingScope(scoped_route_info);
-        return config;
-      });
+      updated_scopes.push_back(scoped_route_info);
       any_applied = true;
       ENVOY_LOG(debug, "srds: add/update scoped_route '{}', version: {}",
                 scoped_route_info->scopeName(), version_info);
     } catch (const EnvoyException& e) {
       exception_msgs.emplace_back(absl::StrCat("", e.what()));
     }
   }
+  if (!updated_scopes.empty()) {
```

Comment:
Note that this changes the logic such that partial application of an SRDS config update is no longer possible. For example, if there is a duplicate scoped route, all preceding routes would have been applied prior to this change.

I actually think the new behavior is preferable; I suggest adding a test to explicitly validate this logic.

FYI @stevenzzzz

Reply:
Yes, currently the srds uses StoW. A test to validate that partial update is not accepted will be good.

Reply:
I think Andres is saying in delta XDS within a response handling, we could have a portion of the list applied, as the previous scope info is distributed to workers. We need to get this behavior tested. 
I actually dont have a strong opinion of it, delta xDS cant send per-resource status back to management server anyway, so go ahead.

Reply:
Even for a regular SOTW config update, the existing logic could partially apply config since `applyConfigUpdate()` is called inline for each scoped route distributed in the `DiscoveryResponse`.

Reply:
for the SotW api, the two errors has been checked before the Delta interface is called. 
Actually a good refactoring could be moving the verification code into a helper fuction. @chaoqin-li1123

Reply:
Good point, thanks Xin. This only applies to delta updates.

Reply:
So if the new change is applied, in delta api, partial updates is accepted in main thread when exception happens, but the partial updates are not applied to worker threads. We want the update to be fully rejected in both main thread and worker thread, right?

Reply:
Maybe we can store all the incoming scoped_route_config in a vector, pass the reference of this vector to a helper function to detect any scope key or scope name conflict.  If there is any conflict, no change will be applied, otherwise all changes will be applied by creating corresponding RdsRouteConfigProviderHelper and push scope_route_info into the maps.

Reply:
A helper function has been added to detect key and scope name conflict.